File: swift/Sources/days/day07.swift
private typealias CardCounts = [Character: Int]

// init CardCounts from cardChars string
private extension Dictionary where Key == Character, Value == Int {
    init(cardChars: String) {
        self.init()
        for char in cardChars {
            self[char, default: 0] += 1
        }
    }
}

private enum HandType: Comparable {
    case HighCard
    case OnePair
    case TwoPair
    case ThreeOfAKind
    case FullHouse
    case FourOfAKind
    case FiveOfAKind

    static func from(cardCounts: CardCounts) -> HandType {
        let maxCount = cardCounts.values.max() ?? 0

        switch (maxCount, cardCounts.count) {
        case (5, _):
            return .FiveOfAKind
        case (4, _):
            return .FourOfAKind
        case (3, 2):
            return .FullHouse
        case (3, _):
            return .ThreeOfAKind
        case (2, 3):
            return .TwoPair
        case (2, _):
            return .OnePair
        default:
            return .HighCard
        }
    }
}

private struct Hand {
    let cardChars: [Character]
    let bid: Int
    let cardCounts: CardCounts
    let type: HandType
}

private enum ParseInputError: Error {
    case unableToParseInput
}

private func parseIntput(_ input: String) throws -> [Hand] {
    let lines = input.split(separator: "\n")
    return try lines.map { line in
        let handData = line.split(separator: " ")
        guard handData.count == 2,
            handData[0].count == 5,
            let bid = Int(handData[1]) else {
            throw ParseInputError.unableToParseInput
        }
        let cardChars = String(handData[0])
        let cardCounts = CardCounts(cardChars: cardChars)
        let handType = HandType.from(cardCounts: cardCounts)
        return Hand(cardChars: Array(cardChars), bid: bid, cardCounts: cardCounts, type: handType)
    }
}

private typealias CardStrengths = [Character: Int]
private typealias HandComparator = (Hand, Hand) -> Bool

private func compareHandStrengths0(cardStrengths: CardStrengths) -> HandComparator {
    return { hand1, hand2 in
        if hand1.type != hand2.type {
            return hand1.type < hand2.type
        }
        for cardIndex in 0..<hand1.cardChars.count {
            if let card1Strength = cardStrengths[hand1.cardChars[cardIndex]],
                let card2Strength = cardStrengths[hand2.cardChars[cardIndex]] {
                if card1Strength != card2Strength {
                    return card1Strength < card2Strength
                }
            }
        }
        return true // we should not have any ties in our data set
    }
}

private let cardOrderPart1: [Character] = ["2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"]
private let cardStrengthsPart1: CardStrengths = Dictionary(uniqueKeysWithValues: zip(cardOrderPart1, 0..<cardOrderPart1.count))
private let compareHandStrengthsPart1 = compareHandStrengths0(cardStrengths: cardStrengthsPart1)

private func part1(input: String) throws -> String {
    let hands = try parseIntput(input)
    let sortedHands = hands.sorted(by: compareHandStrengthsPart1)
    let winnings = sortedHands.enumerated().map { rankIndex, hand in
        let rank = rankIndex + 1
        return rank * hand.bid
    }
    let result = winnings.reduce(0, +)
    return String(result)
}

private func applyWildCards(to hand: Hand) -> Hand {
    guard let jokerCount = hand.cardCounts["J"],
        jokerCount != 0 else {
        return hand
    }

    var newCounts = hand.cardCounts
    newCounts["J"] = nil
    if let maxElement = newCounts.max(by: { $0.value < $1.value }) {
        newCounts[maxElement.key] = maxElement.value + jokerCount
    } else {
        newCounts["A"] = jokerCount // in the event all cards were jokers, we'll replace the hand with all aces
    }
    let newHandType = HandType.from(cardCounts: newCounts)

    return Hand(cardChars: hand.cardChars, bid: hand.bid, cardCounts: newCounts, type: newHandType)
}

private let cardOrderPart2: [Character] = ["J", "2", "3", "4", "5", "6", "7", "8", "9", "T", "Q", "K", "A"]
private let cardStrengthsPart2: CardStrengths = Dictionary(uniqueKeysWithValues: zip(cardOrderPart2, 0..<cardOrderPart2.count))
private let compareHandStrengthsPart2 = compareHandStrengths0(cardStrengths: cardStrengthsPart2)

private func part2(input: String) throws -> String {
    let hands = try parseIntput(input)
    let wildCardsApplied = hands.map(applyWildCards)
    let sortedHands = wildCardsApplied.sorted(by: compareHandStrengthsPart2)
    let winnings = sortedHands.enumerated().map { rankIndex, hand in
        let rank = rankIndex + 1
        return rank * hand.bid
    }
    let result = winnings.reduce(0, +)
    return String(result)
}

func day07(dayKey: String) throws {
    try testAocOutput(dayKey: dayKey, inputName: "example", partKey: "1", partSolver: part1)
    try testAocOutput(dayKey: dayKey, inputName: "test", partKey: "1", partSolver: part1)

    try testAocOutput(dayKey: dayKey, inputName: "example", partKey: "2", partSolver: part2)
    try testAocOutput(dayKey: dayKey, inputName: "test", partKey: "2", partSolver: part2)
}

File: js/src/days/day07.js
import { sum } from "../lib/iterators.js"
import * as runner from "../lib/runner.js"

console.log("Solving AoC 2023 day 07")

const evaluateHandType = (cardCounts) => {
  const maxCount = Math.max.apply(null, [...cardCounts.values()])
  // 5 of a kind
  if (maxCount === 5) { return 6 }
  // 4 of a kind
  if (maxCount === 4) { return 5 }
  // full house
  if (maxCount === 3 && cardCounts.size === 2) { return 4 }
  // 3 of a kind
  if (maxCount === 3 && cardCounts.size > 2) { return 3 }
  // two pair
  if (maxCount === 2 && cardCounts.size === 3) { return 2 }
  // one pair
  if (maxCount === 2 && cardCounts.size > 3) { return 1 }
  // high card
  return 0
}

const parseInput = (input) => {
  const lines = input.split('\n')
  return lines.map((line) => {
    const [handString, bidStr] = line.split(" ")
    const bid = Number(bidStr)
    const cardCounts = new Map()
    for (const cardChar of handString) {
      const currentCount = cardCounts.get(cardChar) || 0
      cardCounts.set(cardChar, currentCount + 1)
    }
    const handType = evaluateHandType(cardCounts)
    return {handString, bid, cardCounts, handType}
  })
}

const compareHandStrengths0 = (cardStrengths) => (hand1, hand2) => {
  const typeDifference = hand1.handType - hand2.handType
  if (typeDifference !== 0) { return typeDifference }
  for (let i = 0; i < hand1.handString.length; i++) {
    const card1Strength = cardStrengths[hand1.handString[i]]
    const card2Strength = cardStrengths[hand2.handString[i]]
    if (card1Strength > card2Strength) { return 1 }
    if (card1Strength < card2Strength) { return -1 }
  }
  console.error("its a tie!", hand1.handString === hand2.handString, hand1.handString, hand2.handString)
  return 0
}

const cardStrengthsOrderPart1 = ["2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"]
const cardStrengthsPart1 = cardStrengthsOrderPart1.reduce((acc, cardChar, i) => {
  acc[cardChar] = i
  return acc
}, {})

const compareHandStrengthsPart1 = compareHandStrengths0(cardStrengthsPart1)

const part1 = (input) => {
  const hands = parseInput(input)
  const sortedHands = hands.toSorted(compareHandStrengthsPart1)
  const winnings = sortedHands.map((hand, rankIndex) => {
    const rank = rankIndex + 1
    return hand.bid * rank
  })
  return sum(winnings)
}

const cardStrengthsOrderPart2 = ["J", "2", "3", "4", "5", "6", "7", "8", "9", "T", "Q", "K", "A"]
const cardStrengthsPart2 = cardStrengthsOrderPart2.reduce((acc, cardChar, i) => {
  acc[cardChar] = i
  return acc
}, {})

const compareHandStrengthsPart2 = compareHandStrengths0(cardStrengthsPart2)

// mutates hand cardCounts and cardType in place
const applyWildCards = (hand) => {
  const jokerCount = hand.cardCounts.get("J") || 0
  if (jokerCount === 0) { return }

  const newCounts = new Map()
  let maxCardCount = 0
  let maxCardChar = "A" // in the event of 5 jokers, we'll end up with a hand of 5 aces
  for (const [cardChar, cardCount] of hand.cardCounts) {
    if (cardChar === "J") { continue }
    newCounts.set(cardChar, cardCount)
    if (cardCount > maxCardCount) {
      maxCardCount = cardCount
      maxCardChar = cardChar
    }
  }
  newCounts.set(maxCardChar, maxCardCount + jokerCount) // add the joker count to whichever card currently has the highest count

  hand.cardCounts = newCounts
  hand.handType = evaluateHandType(newCounts)
}

const part2 = (input) => {
  const hands = parseInput(input)
  hands.forEach(applyWildCards)
  const sortedHands = hands.toSorted(compareHandStrengthsPart2)
  const winnings = sortedHands.map((hand, rankIndex) => {
    const rank = rankIndex + 1
    return hand.bid * rank
  })
  return sum(winnings)
}

await runner.testOutput('day07/example', '1', part1)
// await runner.printOutput('day07/test', part1)
// await runner.copyOutput('day07/test', part1)
// await runner.writeOutput('day07/test', '1', part1)
await runner.testOutput('day07/test', '1', part1)

await runner.testOutput('day07/example', '2', part2)
// await runner.printOutput('day07/test', part2)
// await runner.copyOutput('day07/test', part2)
// await runner.writeOutput('day07/test', '2', part2)
await runner.testOutput('day07/test', '2', part2)
